spec/controllers/books_controller_spec.rb

```diff
@@ -0,0 +1,144 @@
+require 'rails_helper'
+
+RSpec.describe BooksController, type: :controller do
+  describe 'GET /index' do
+    subject { get :index }
+
+    it 'returns all books' do
+      expect(subject.status).to eq(200)
+      expect(assigns(:books)).to eq(Book.all)
+    end
+  end
+
+  describe 'GET /show' do
+    subject { get :show, params: }
+
+    let(:book) { create(:book) }
+    let(:params) { { id: book.id } }
+
+    context 'when book exist' do
+      it 'returns book' do
+        expect(subject.status).to eq(200)
+        expect(subject).to render_template(:show)
+        expect(assigns(:book)).to eq(Book.find(book.id))
+      end
+    end
+
+    context 'when book not exist' do
+      let(:params) { { id: -1 } }
+
+      it 'raise error not found' do
+        # expect(subject.status).to eq(404)
+        expect { subject }.to raise_error(ActiveRecord::RecordNotFound)
+      end
+    end
+  end
+
+  describe 'GET /new' do
+    subject { get :new }
+
+    it 'renders to new book form' do
+      expect(subject.status).to eq(200)
+      expect(subject).to render_template(:new)
+    end
+  end
+
+  describe 'POST /create' do
+    subject { post :create, params: }
+
+    let(:params) { { book: { name: 'Controller book', description: 'Book from controller', release: '2023-01-01' } } }
+
+    it 'creates a book success' do
+      expect(subject.status).to eq(302)
+      expect(Book.count).to eq(1)
+
+      # test value of each attributes
+      created_book = assigns(:book)
+      expect(Book.find(created_book.id).name).to eq('Controller book')
+      expect(Book.find(created_book.id).description).to eq('Book from controller')
+      expect(Book.find(created_book.id).release).to eq(Date.new(2023, 1, 1))
```

Comment:
```suggestion
      expect(Book.find(created_book.id).as_json).to hash_including (
        'name' => 'Controller book'
        'description' => 'Book from controller'
        'release' => Date.new(2023, 1, 1)
      )
```
Do this look better? 😂  If not maybe we can just define `Book.find(created_book.id)` as a variable
Not sure if it working pls recheck krub

Reply:
ลองแล้วมันขึ้น ยังงี้แทนอะค่ะ 🥹
```
NoMethodError:
       undefined method `matches?' for #<RSpec::Mocks::ArgumentMatchers::HashIncludingMatcher:0x0000000109c6d8d8 @expected={"name"=>"Controller book", "description"=>"Book from controller", "release"=>Sun, 01 Jan 2023}>
```

Reply:
`expect(Book.find(created_book.id).as_json).to match hash_including`
ถ้าเพิ่ม match ไปหลัง .to ได้มั้ยนะครับ

อีกอย่างคือไม่แน่ใจว่าข้างในต้องเป็น `'name' => '...'` หรือ `name: '...'`
ลองเล่นดูก็ได้ครับ หรือถ้าไม่เวิร์ค เอาเป็น

```suggestion
      book = Book.find(created_book.id)
      expect(book.name).to eq('Controller book')
      expect(book.description).to eq('Book from controller')
      expect(book.release).to eq(Date.new(2023, 1, 1))
```

ก็ได้ครับ